# Sorting, Counting, Subsetting

## Import libraries

In [140]:
## import libraries
import pandas as pd

## Upload Olympics datasets to this notebook.

### Import ```olympics-summer.csv``` and store in a dataframe called ```dfs``` (<a href="https://raw.githubusercontent.com/sandeepmj/datasets/main/olympics-summer.csv">link</a>).


### Import ```olympics-winter.csv``` and store in a dataframe called ```dfw``` (<a href="https://raw.githubusercontent.com/sandeepmj/datasets/main/olympics-winter.csv">link</a>).

In [141]:
## import summer data
dfs = pd.read_csv("https://raw.githubusercontent.com/sandeepmj/datasets/main/olympics-summer.csv")
(dfs)

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal
0,1724,Aristidis Akratopoulos,M,NaN,NaN,NaN,Greece,GRE,1896,Athina,Tennis,Tennis Men's Singles,NaN
1,1724,Aristidis Akratopoulos,M,NaN,NaN,NaN,Greece-3,GRE,1896,Athina,Tennis,Tennis Men's Doubles,NaN
2,1725,"Konstantinos ""Kostas"" Akratopoulos",M,NaN,NaN,NaN,Greece,GRE,1896,Athina,Tennis,Tennis Men's Singles,NaN
3,1725,"Konstantinos ""Kostas"" Akratopoulos",M,NaN,NaN,NaN,Greece-3,GRE,1896,Athina,Tennis,Tennis Men's Doubles,NaN
4,4113,Anastasios Andreou,M,NaN,NaN,NaN,Greece,GRE,1896,Athina,Athletics,Athletics Men's 110 metres Hurdles,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...
222547,135528,Marc Zwiebler,M,32.0,181.0,75.0,Germany,GER,2016,Rio de Janeiro,Badminton,Badminton Men's Singles,NaN
222548,135547,Viktoriya Viktorovna Zyabkina,F,23.0,174.0,62.0,Kazakhstan,KAZ,2016,Rio de Janeiro,Athletics,Athletics Women's 100 metres,NaN
222549,135547,Viktoriya Viktorovna Zyabkina,F,23.0,174.0,62.0,Kazakhstan,KAZ,2016,Rio de Janeiro,Athletics,Athletics Women's 200 metres,NaN
222550,135547,Viktoriya Viktorovna Zyabkina,F,23.0,174.0,62.0,Kazakhstan,KAZ,2016,Rio de Janeiro,Athletics,Athletics Women's 4 x 100 metres Relay,NaN


In [142]:
## import winter data
dfw = pd.read_csv("https://raw.githubusercontent.com/sandeepmj/datasets/main/olympics-winter.csv")
(dfw)

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal
0,391,Clarence John Abel,M,23.0,185.0,102.0,United States,USA,1924,Chamonix,Ice Hockey,Ice Hockey Men's Ice Hockey,Silver
1,992,Josef Adolf,M,25.0,NaN,NaN,Czechoslovakia,TCH,1924,Chamonix,Nordic Combined,Nordic Combined Men's Individual,NaN
2,1077,Xavier Affentranger,M,26.0,NaN,NaN,Switzerland,SUI,1924,Chamonix,Ski Jumping,"Ski Jumping Men's Normal Hill, Individual",NaN
3,1077,Xavier Affentranger,M,26.0,NaN,NaN,Switzerland,SUI,1924,Chamonix,Cross Country Skiing,Cross Country Skiing Men's 18 kilometres,NaN
4,1077,Xavier Affentranger,M,26.0,NaN,NaN,Switzerland,SUI,1924,Chamonix,Nordic Combined,Nordic Combined Men's Individual,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...
48559,135485,Stepan Olegovich Zuyev,M,25.0,189.0,90.0,Russia,RUS,2014,Sochi,Alpine Skiing,Alpine Skiing Men's Slalom,NaN
48560,135500,Kristaps Zvejnieks,M,21.0,185.0,87.0,Latvia,LAT,2014,Sochi,Alpine Skiing,Alpine Skiing Men's Giant Slalom,NaN
48561,135500,Kristaps Zvejnieks,M,21.0,185.0,87.0,Latvia,LAT,2014,Sochi,Alpine Skiing,Alpine Skiing Men's Slalom,NaN
48562,135570,Piotr ya,M,27.0,176.0,59.0,Poland,POL,2014,Sochi,Ski Jumping,"Ski Jumping Men's Large Hill, Individual",NaN


## Get a sense of the data

In [143]:
## SUMMER
## what exactly do we have: columns, datatypes, etc.
dfs.info()

<class 'pandas.DataFrame'>
RangeIndex: 222552 entries, 0 to 222551
Data columns (total 13 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   ID            222552 non-null  int64  
 1   Athlete Name  222552 non-null  str    
 2   Sex           222552 non-null  str    
 3   Age           213363 non-null  float64
 4   Height        170695 non-null  float64
 5   Weight        168698 non-null  float64
 6   Team          222552 non-null  str    
 7   NOC           222552 non-null  str    
 8   Year          222552 non-null  int64  
 9   City          222552 non-null  str    
 10  Sport         222552 non-null  str    
 11  Event         222552 non-null  str    
 12  Medal         34088 non-null   str    
dtypes: float64(3), int64(2), str(8)
memory usage: 22.1 MB


In [144]:
## WINTER
## what exactly do we have: columns, datatypes, etc.
dfw.info()

<class 'pandas.DataFrame'>
RangeIndex: 48564 entries, 0 to 48563
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ID            48564 non-null  int64  
 1   Athlete Name  48564 non-null  str    
 2   Sex           48564 non-null  str    
 3   Age           48279 non-null  float64
 4   Height        40250 non-null  float64
 5   Weight        39543 non-null  float64
 6   Team          48564 non-null  str    
 7   NOC           48564 non-null  str    
 8   Year          48564 non-null  int64  
 9   City          48564 non-null  str    
 10  Sport         48564 non-null  str    
 11  Event         48564 non-null  str    
 12  Medal         5695 non-null   str    
dtypes: float64(3), int64(2), str(8)
memory usage: 4.8 MB


## Subsetting dataframes into smaller, targeted dataframes

When you subset a dataframe, you are filtering down based on selecting specific columns. We want to winnow a large dataset to manageable size by including only relevant columns.

There are two ways of doing this. One simple method is very common (you'll see many people use it), but the other method (also simple) is way more powerful and flexible.

### Method 1 - Pass a list of multiple column headers, it returns dataframe with only those columns
This is like calling a series (```df["Year"]```), but multiple columns:

####  ```df[["Column 1", "Column 2"]]```

In [145]:
## Create a dataframe that for only summer games years and cities
dfs.info()

dfs_yr_city= dfs[["Year", "City"]]

<class 'pandas.DataFrame'>
RangeIndex: 222552 entries, 0 to 222551
Data columns (total 13 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   ID            222552 non-null  int64  
 1   Athlete Name  222552 non-null  str    
 2   Sex           222552 non-null  str    
 3   Age           213363 non-null  float64
 4   Height        170695 non-null  float64
 5   Weight        168698 non-null  float64
 6   Team          222552 non-null  str    
 7   NOC           222552 non-null  str    
 8   Year          222552 non-null  int64  
 9   City          222552 non-null  str    
 10  Sport         222552 non-null  str    
 11  Event         222552 non-null  str    
 12  Medal         34088 non-null   str    
dtypes: float64(3), int64(2), str(8)
memory usage: 22.1 MB


## Sorting by single value

```dataframe.sort_values(by="Value 1", ascending = True)``` for A-Z or smaller numbers to bigger numbers

```ascending = False``` for Z-A or bigger to smaller numbers.

In [146]:
## sort by athlete names A-Z for winter olympics

dfw.sort_values(by="Athlete Name", ascending = True)

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal
2788,56427,Aage Justesen,M,30.0,NaN,NaN,Denmark,DEN,1948,Sankt Moritz,Speed Skating,Speed Skating Men's 500 metres,NaN
15974,13348,"Aaltje Grietje ""Alie"" Boorsma",F,24.0,179.0,75.0,Netherlands,NED,1984,Sarajevo,Speed Skating,Speed Skating Women's 500 metres,NaN
15975,13348,"Aaltje Grietje ""Alie"" Boorsma",F,24.0,179.0,75.0,Netherlands,NED,1984,Sarajevo,Speed Skating,"Speed Skating Women's 1,000 metres",NaN
3799,49758,Aarne Vin Edward Honkavaara,M,27.0,184.0,81.0,Finland,FIN,1952,Oslo,Ice Hockey,Ice Hockey Men's Ice Hockey,NaN
44053,12407,Aaron Blunck,M,17.0,180.0,78.0,United States,USA,2014,Sochi,Freestyle Skiing,Freestyle Skiing Men's Halfpipe,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...
15622,122902,yvind Tveter,M,25.0,190.0,89.0,Norway,NOR,1980,Lake Placid,Speed Skating,"Speed Skating Men's 5,000 metres",NaN
15623,122902,yvind Tveter,M,25.0,190.0,89.0,Norway,NOR,1980,Lake Placid,Speed Skating,"Speed Skating Men's 10,000 metres",NaN
8706,5787,zer Atei,M,25.0,180.0,72.0,Turkey,TUR,1968,Grenoble,Alpine Skiing,Alpine Skiing Men's Slalom,NaN
8705,5787,zer Atei,M,25.0,180.0,72.0,Turkey,TUR,1968,Grenoble,Alpine Skiing,Alpine Skiing Men's Giant Slalom,NaN


In [147]:
## sort city z-a

dfw.sort_values(by="City", ascending = False)

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal
39561,11951,Ole Einar Bjrndalen,M,36.0,178.0,65.0,Norway,NOR,2010,Vancouver,Biathlon,Biathlon Men's 4 x 7.5 kilometres Relay,Gold
40747,52463,Marc Iselin,M,29.0,178.0,83.0,Switzerland,SUI,2010,Vancouver,Snowboarding,Snowboarding Men's Parallel Giant Slalom,NaN
40741,52316,Dmitry Gennadyevich Ipatov,M,25.0,176.0,60.0,Russia,RUS,2010,Vancouver,Ski Jumping,"Ski Jumping Men's Large Hill, Team",NaN
40742,52345,"Michael James Grant ""Mike"" Ireland",M,36.0,181.0,79.0,Canada,CAN,2010,Vancouver,Speed Skating,Speed Skating Men's 500 metres,NaN
40743,52392,Haley Lyn Irwin,F,21.0,170.0,78.0,Canada,CAN,2010,Vancouver,Ice Hockey,Ice Hockey Women's Ice Hockey,Gold
...,...,...,...,...,...,...,...,...,...,...,...,...,...
22495,75068,Dennis Edward Marineau,M,29.0,172.0,81.0,Canada-2,CAN,1992,Albertville,Bobsleigh,Bobsleigh Men's Two,NaN
22494,74979,Diego Margozzini,M,21.0,182.0,76.0,Chile,CHI,1992,Albertville,Alpine Skiing,Alpine Skiing Men's Giant Slalom,NaN
22493,74960,Pascal Margerit,M,20.0,180.0,75.0,France,FRA,1992,Albertville,Ice Hockey,Ice Hockey Men's Ice Hockey,NaN
22492,74931,Alexander Marent,M,22.0,168.0,68.0,Austria,AUT,1992,Albertville,Cross Country Skiing,Cross Country Skiing Men's 4 x 10 kilometres R...,NaN


## Sorting by multiple values

```dataframe.sort_values(by=["Value 1", "Value 2"], ascending = True)``` for A-Z or smaller numbers to bigger numbers

```ascending = False``` for Z-A or bigger to smaller numbers.


In [148]:
## Sort by values (City and Year) for winter olympics
dfw.sort_values(by =["City", "Year"], ascending = True)

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal
20579,5,Christine Jacoba Aaftink,F,25.0,185.0,82.0,Netherlands,NED,1992,Albertville,Speed Skating,Speed Skating Women's 500 metres,NaN
20580,5,Christine Jacoba Aaftink,F,25.0,185.0,82.0,Netherlands,NED,1992,Albertville,Speed Skating,"Speed Skating Women's 1,000 metres",NaN
20581,6,Per Knut Aaland,M,31.0,188.0,75.0,United States,USA,1992,Albertville,Cross Country Skiing,Cross Country Skiing Men's 10 kilometres,NaN
20582,6,Per Knut Aaland,M,31.0,188.0,75.0,United States,USA,1992,Albertville,Cross Country Skiing,Cross Country Skiing Men's 50 kilometres,NaN
20583,6,Per Knut Aaland,M,31.0,188.0,75.0,United States,USA,1992,Albertville,Cross Country Skiing,Cross Country Skiing Men's 10/15 kilometres Pu...,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...
43668,135492,Veronika Zuzulov (-Velez),F,25.0,172.0,72.0,Slovakia,SVK,2010,Vancouver,Alpine Skiing,Alpine Skiing Women's Giant Slalom,NaN
43669,135492,Veronika Zuzulov (-Velez),F,25.0,172.0,72.0,Slovakia,SVK,2010,Vancouver,Alpine Skiing,Alpine Skiing Women's Slalom,NaN
43670,135496,Veronika Zvaiov,F,21.0,176.0,62.0,Czech Republic,CZE,2010,Vancouver,Biathlon,Biathlon Women's 7.5 kilometres Sprint,NaN
43671,135500,Kristaps Zvejnieks,M,17.0,185.0,87.0,Latvia,LAT,2010,Vancouver,Alpine Skiing,Alpine Skiing Men's Giant Slalom,NaN


In [149]:
## sort by year and then city
dfw_year_city_sort = dfw.sort_values(by =["Year", "City"], ascending = True)
dfw_year_city_sort.head(10)

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal
0,391,Clarence John Abel,M,23.0,185.0,102.0,United States,USA,1924,Chamonix,Ice Hockey,Ice Hockey Men's Ice Hockey,Silver
1,992,Josef Adolf,M,25.0,NaN,NaN,Czechoslovakia,TCH,1924,Chamonix,Nordic Combined,Nordic Combined Men's Individual,NaN
2,1077,Xavier Affentranger,M,26.0,NaN,NaN,Switzerland,SUI,1924,Chamonix,Ski Jumping,"Ski Jumping Men's Normal Hill, Individual",NaN
3,1077,Xavier Affentranger,M,26.0,NaN,NaN,Switzerland,SUI,1924,Chamonix,Cross Country Skiing,Cross Country Skiing Men's 18 kilometres,NaN
4,1077,Xavier Affentranger,M,26.0,NaN,NaN,Switzerland,SUI,1924,Chamonix,Nordic Combined,Nordic Combined Men's Individual,NaN
5,1341,Johan Petter hln (Andersson-),M,44.0,NaN,NaN,Sweden,SWE,1924,Chamonix,Curling,Curling Men's Curling,Silver
6,2329,Louis Albert,M,25.0,NaN,NaN,France,FRA,1924,Chamonix,Ski Jumping,"Ski Jumping Men's Normal Hill, Individual",NaN
7,2431,Henri Eugne Aldebert,M,43.0,NaN,NaN,France-1,FRA,1924,Chamonix,Bobsleigh,Bobsleigh Men's Four/Five,NaN
8,2902,Karl Ruben Allinger,M,32.0,NaN,NaN,Sweden,SWE,1924,Chamonix,Ice Hockey,Ice Hockey Men's Ice Hockey,NaN
9,2939,Ernst Alm,M,23.0,174.0,NaN,Sweden,SWE,1924,Chamonix,Cross Country Skiing,Cross Country Skiing Men's 50 kilometres,NaN


In [150]:
## sort by city and year but descending for summer games
dfs_city_year_sort = dfs.sort_values(by = ["City", "Year"], ascending = True)
dfs_city_year_sort.head(20)

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal
22016,47,Erling Rudolf Aastad,M,30.0,177.0,74.0,Norway,NOR,1928,Amsterdam,Athletics,Athletics Men's Long Jump,NaN
22017,102,Sayed Fahmy Abaza,M,NaN,NaN,NaN,Egypt,EGY,1928,Amsterdam,Football,Football Men's Football,NaN
22018,142,"David ""Dave"" Abbott",M,26.0,183.0,75.0,United States,USA,1928,Amsterdam,Athletics,"Athletics Men's 5,000 metres",NaN
22019,255,Mahmoud Ahmed Abdin,M,21.0,NaN,NaN,Egypt,EGY,1928,Amsterdam,Fencing,"Fencing Men's Foil, Individual",NaN
22020,255,Mahmoud Ahmed Abdin,M,21.0,NaN,NaN,Egypt,EGY,1928,Amsterdam,Fencing,"Fencing Men's Foil, Team",NaN
22021,281,S. Abdul Hamid,M,NaN,NaN,NaN,India,IND,1928,Amsterdam,Athletics,Athletics Men's 110 metres Hurdles,NaN
22022,281,S. Abdul Hamid,M,NaN,NaN,NaN,India,IND,1928,Amsterdam,Athletics,Athletics Men's 400 metres Hurdles,NaN
22023,385,"Max ""Xam"" Abegglen",M,26.0,NaN,NaN,Switzerland,SUI,1928,Amsterdam,Football,Football Men's Football,NaN
22024,389,Adolf Gaston Abel,M,45.0,NaN,NaN,Germany,GER,1928,Amsterdam,Art Competitions,"Art Competitions Mixed Architecture, Designs F...",NaN
22025,389,Adolf Gaston Abel,M,45.0,NaN,NaN,Germany,GER,1928,Amsterdam,Art Competitions,"Art Competitions Mixed Architecture, Architect...",NaN


In [151]:
## you can store into a new df


## Please don't open till instructed

### Should Winter and Summer data remain separate?

### What if we want to know total gold medals for any country?

### But if we combine them, how to we easily distinguish between summer and winter?

### Creating a new column with a default value

In [152]:
## In the summer dataframe, add "Season" column with the value "Summer"

dfs["Season"] = "summer"
dfs


,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal,Season
0,1724,Aristidis Akratopoulos,M,NaN,NaN,NaN,Greece,GRE,1896,Athina,Tennis,Tennis Men's Singles,NaN,summer
1,1724,Aristidis Akratopoulos,M,NaN,NaN,NaN,Greece-3,GRE,1896,Athina,Tennis,Tennis Men's Doubles,NaN,summer
2,1725,"Konstantinos ""Kostas"" Akratopoulos",M,NaN,NaN,NaN,Greece,GRE,1896,Athina,Tennis,Tennis Men's Singles,NaN,summer
3,1725,"Konstantinos ""Kostas"" Akratopoulos",M,NaN,NaN,NaN,Greece-3,GRE,1896,Athina,Tennis,Tennis Men's Doubles,NaN,summer
4,4113,Anastasios Andreou,M,NaN,NaN,NaN,Greece,GRE,1896,Athina,Athletics,Athletics Men's 110 metres Hurdles,NaN,summer
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
222547,135528,Marc Zwiebler,M,32.0,181.0,75.0,Germany,GER,2016,Rio de Janeiro,Badminton,Badminton Men's Singles,NaN,summer
222548,135547,Viktoriya Viktorovna Zyabkina,F,23.0,174.0,62.0,Kazakhstan,KAZ,2016,Rio de Janeiro,Athletics,Athletics Women's 100 metres,NaN,summer
222549,135547,Viktoriya Viktorovna Zyabkina,F,23.0,174.0,62.0,Kazakhstan,KAZ,2016,Rio de Janeiro,Athletics,Athletics Women's 200 metres,NaN,summer
222550,135547,Viktoriya Viktorovna Zyabkina,F,23.0,174.0,62.0,Kazakhstan,KAZ,2016,Rio de Janeiro,Athletics,Athletics Women's 4 x 100 metres Relay,NaN,summer


In [153]:
## In the winter dataframe, add "Season" column with the value "Winter"

dfw["Season"] = "winter"
dfw

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal,Season
0,391,Clarence John Abel,M,23.0,185.0,102.0,United States,USA,1924,Chamonix,Ice Hockey,Ice Hockey Men's Ice Hockey,Silver,winter
1,992,Josef Adolf,M,25.0,NaN,NaN,Czechoslovakia,TCH,1924,Chamonix,Nordic Combined,Nordic Combined Men's Individual,NaN,winter
2,1077,Xavier Affentranger,M,26.0,NaN,NaN,Switzerland,SUI,1924,Chamonix,Ski Jumping,"Ski Jumping Men's Normal Hill, Individual",NaN,winter
3,1077,Xavier Affentranger,M,26.0,NaN,NaN,Switzerland,SUI,1924,Chamonix,Cross Country Skiing,Cross Country Skiing Men's 18 kilometres,NaN,winter
4,1077,Xavier Affentranger,M,26.0,NaN,NaN,Switzerland,SUI,1924,Chamonix,Nordic Combined,Nordic Combined Men's Individual,NaN,winter
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
48559,135485,Stepan Olegovich Zuyev,M,25.0,189.0,90.0,Russia,RUS,2014,Sochi,Alpine Skiing,Alpine Skiing Men's Slalom,NaN,winter
48560,135500,Kristaps Zvejnieks,M,21.0,185.0,87.0,Latvia,LAT,2014,Sochi,Alpine Skiing,Alpine Skiing Men's Giant Slalom,NaN,winter
48561,135500,Kristaps Zvejnieks,M,21.0,185.0,87.0,Latvia,LAT,2014,Sochi,Alpine Skiing,Alpine Skiing Men's Slalom,NaN,winter
48562,135570,Piotr ya,M,27.0,176.0,59.0,Poland,POL,2014,Sochi,Ski Jumping,"Ski Jumping Men's Large Hill, Individual",NaN,winter


In [154]:
## are the two column headers identical?
dfw.info()

<class 'pandas.DataFrame'>
RangeIndex: 48564 entries, 0 to 48563
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ID            48564 non-null  int64  
 1   Athlete Name  48564 non-null  str    
 2   Sex           48564 non-null  str    
 3   Age           48279 non-null  float64
 4   Height        40250 non-null  float64
 5   Weight        39543 non-null  float64
 6   Team          48564 non-null  str    
 7   NOC           48564 non-null  str    
 8   Year          48564 non-null  int64  
 9   City          48564 non-null  str    
 10  Sport         48564 non-null  str    
 11  Event         48564 non-null  str    
 12  Medal         5695 non-null   str    
 13  Season        48564 non-null  str    
dtypes: float64(3), int64(2), str(9)
memory usage: 5.2 MB


In [155]:
dfs.info()

<class 'pandas.DataFrame'>
RangeIndex: 222552 entries, 0 to 222551
Data columns (total 14 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   ID            222552 non-null  int64  
 1   Athlete Name  222552 non-null  str    
 2   Sex           222552 non-null  str    
 3   Age           213363 non-null  float64
 4   Height        170695 non-null  float64
 5   Weight        168698 non-null  float64
 6   Team          222552 non-null  str    
 7   NOC           222552 non-null  str    
 8   Year          222552 non-null  int64  
 9   City          222552 non-null  str    
 10  Sport         222552 non-null  str    
 11  Event         222552 non-null  str    
 12  Medal         34088 non-null   str    
 13  Season        222552 non-null  str    
dtypes: float64(3), int64(2), str(9)
memory usage: 23.8 MB


In [156]:
## join winter and summer dataframes together into a dataframe called df
df = pd.concat([dfw, dfs], ignore_index = True)
(df)

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal,Season
0,391,Clarence John Abel,M,23.0,185.0,102.0,United States,USA,1924,Chamonix,Ice Hockey,Ice Hockey Men's Ice Hockey,Silver,winter
1,992,Josef Adolf,M,25.0,NaN,NaN,Czechoslovakia,TCH,1924,Chamonix,Nordic Combined,Nordic Combined Men's Individual,NaN,winter
2,1077,Xavier Affentranger,M,26.0,NaN,NaN,Switzerland,SUI,1924,Chamonix,Ski Jumping,"Ski Jumping Men's Normal Hill, Individual",NaN,winter
3,1077,Xavier Affentranger,M,26.0,NaN,NaN,Switzerland,SUI,1924,Chamonix,Cross Country Skiing,Cross Country Skiing Men's 18 kilometres,NaN,winter
4,1077,Xavier Affentranger,M,26.0,NaN,NaN,Switzerland,SUI,1924,Chamonix,Nordic Combined,Nordic Combined Men's Individual,NaN,winter
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
271111,135528,Marc Zwiebler,M,32.0,181.0,75.0,Germany,GER,2016,Rio de Janeiro,Badminton,Badminton Men's Singles,NaN,summer
271112,135547,Viktoriya Viktorovna Zyabkina,F,23.0,174.0,62.0,Kazakhstan,KAZ,2016,Rio de Janeiro,Athletics,Athletics Women's 100 metres,NaN,summer
271113,135547,Viktoriya Viktorovna Zyabkina,F,23.0,174.0,62.0,Kazakhstan,KAZ,2016,Rio de Janeiro,Athletics,Athletics Women's 200 metres,NaN,summer
271114,135547,Viktoriya Viktorovna Zyabkina,F,23.0,174.0,62.0,Kazakhstan,KAZ,2016,Rio de Janeiro,Athletics,Athletics Women's 4 x 100 metres Relay,NaN,summer


In [157]:
### confirm that you have the correct number of rows after the join.
### you can scroll up to count the number of entries in oly_w and oly_s.
### The totals should be equal to the value generated below


## summer + winter row totals
222552+48564 

271116

## Improving memory allocation

If we were working with a massive dataset, we'd want to find ways to improve processing power.

For example, the following columns have only a few values that repeat again and again as strings. This is highly inefficient.

- The "Season" column has "winter" and "summer",
- The "Medal" column only has "gold", "silver" and "bronze",
- "Gender" at this point in the Olympics only has "male" and "female".

One way to improve memory allocation is to take columns with that contain only a few data points and turn them into categories.



In [158]:
## you should see how "Season", "Medal" and "Gender" are all string objects.
## Note that memory usage: 29+ MB


### Convert to single column to category
```df["column_name"].astype("category")```

In [159]:
## convert gender to category


In [160]:
## INFO

df["Sex"]= df["Sex"].astype('category')
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 271116 entries, 0 to 271115
Data columns (total 14 columns):
 #   Column        Non-Null Count   Dtype   
---  ------        --------------   -----   
 0   ID            271116 non-null  int64   
 1   Athlete Name  271116 non-null  str     
 2   Sex           271116 non-null  category
 3   Age           261642 non-null  float64 
 4   Height        210945 non-null  float64 
 5   Weight        208241 non-null  float64 
 6   Team          271116 non-null  str     
 7   NOC           271116 non-null  str     
 8   Year          271116 non-null  int64   
 9   City          271116 non-null  str     
 10  Sport         271116 non-null  str     
 11  Event         271116 non-null  str     
 12  Medal         39783 non-null   str     
 13  Season        271116 non-null  str     
dtypes: category(1), float64(3), int64(2), str(8)
memory usage: 27.1 MB


### Convert to multiple columns to category

 ```df[["column_name1", "column_name2]].astype("category")```

In [161]:
## convert medal and season  to catagory
df[["Medal", "Season"]] = df[["Medal","Season"]].astype('category')

In [162]:
## See what the memory allocation is now using info()
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 271116 entries, 0 to 271115
Data columns (total 14 columns):
 #   Column        Non-Null Count   Dtype   
---  ------        --------------   -----   
 0   ID            271116 non-null  int64   
 1   Athlete Name  271116 non-null  str     
 2   Sex           271116 non-null  category
 3   Age           261642 non-null  float64 
 4   Height        210945 non-null  float64 
 5   Weight        208241 non-null  float64 
 6   Team          271116 non-null  str     
 7   NOC           271116 non-null  str     
 8   Year          271116 non-null  int64   
 9   City          271116 non-null  str     
 10  Sport         271116 non-null  str     
 11  Event         271116 non-null  str     
 12  Medal         39783 non-null   category
 13  Season        271116 non-null  category
dtypes: category(3), float64(3), int64(2), str(6)
memory usage: 23.5 MB


# It should have dropped to 23.5MB.
That's a big descrease for just about 270,000 rows.

This would have an even bigger impact on larger datasets.

# Query to subset

Create smaller datasets with only the information you want



### Create a subset that holds the results from the Sochi Olympics

```df.query("ColumnName == 'some_value'")```

In [163]:
## return df that holds only Sochi data

df.query("City == 'Sochi'")

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal,Season
43673,16,Juhamatti Tapio Aaltonen,M,28.0,184.0,85.0,Finland,FIN,2014,Sochi,Ice Hockey,Ice Hockey Men's Ice Hockey,Bronze,winter
43674,126,Forough Abbasi,F,20.0,164.0,58.0,Iran,IRI,2014,Sochi,Alpine Skiing,Alpine Skiing Women's Slalom,NaN,winter
43675,145,Jeremy Abbott,M,28.0,175.0,70.0,United States,USA,2014,Sochi,Figure Skating,Figure Skating Men's Singles,NaN,winter
43676,145,Jeremy Abbott,M,28.0,175.0,70.0,United States,USA,2014,Sochi,Figure Skating,Figure Skating Mixed Team,Bronze,winter
43677,463,Agnese boltia,F,17.0,166.0,60.0,Latvia,LAT,2014,Sochi,Alpine Skiing,Alpine Skiing Women's Super G,NaN,winter
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
48559,135485,Stepan Olegovich Zuyev,M,25.0,189.0,90.0,Russia,RUS,2014,Sochi,Alpine Skiing,Alpine Skiing Men's Slalom,NaN,winter
48560,135500,Kristaps Zvejnieks,M,21.0,185.0,87.0,Latvia,LAT,2014,Sochi,Alpine Skiing,Alpine Skiing Men's Giant Slalom,NaN,winter
48561,135500,Kristaps Zvejnieks,M,21.0,185.0,87.0,Latvia,LAT,2014,Sochi,Alpine Skiing,Alpine Skiing Men's Slalom,NaN,winter
48562,135570,Piotr ya,M,27.0,176.0,59.0,Poland,POL,2014,Sochi,Ski Jumping,"Ski Jumping Men's Large Hill, Individual",NaN,winter


### Note the order of the single and double quotes:

If you open with double quotes, the inner quotes have to be single quotes:

```df.query("City == 'Sochi'")```

If you open with single quotes, the inner quotes have to be double quotes:

```df.query('City == "Sochi"')```


Using the same outer and inner quotes will break your code:
</br>
```df.query("City == "Sochi"") ## will break ```

In [164]:
## Store subset into dataframe called sochi_df
sochi_df = df.query("City == 'Sochi'")

**Syntax for querying multi-word column names:**

```dataframe.query('`column name` comparison_operator value')```

Note the opening and closing tick marks ``` ` ```

In [165]:
df.to_csv('all-olympics.csv', index=False)

In [166]:
#will break


In [167]:
## query to return on french athletes
df.query("Team == 'France'")

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal,Season
6,2329,Louis Albert,M,25.0,NaN,NaN,France,FRA,1924,Chamonix,Ski Jumping,"Ski Jumping Men's Normal Hill, Individual",NaN,winter
11,4060,Georges Jules Andr,M,47.0,NaN,NaN,France,FRA,1924,Chamonix,Curling,Curling Men's Curling,Bronze,winter
19,7343,Klber Balmat,M,27.0,NaN,NaN,France,FRA,1924,Chamonix,Ski Jumping,"Ski Jumping Men's Normal Hill, Individual",NaN,winter
20,7343,Klber Balmat,M,27.0,NaN,NaN,France,FRA,1924,Chamonix,Nordic Combined,Nordic Combined Men's Individual,NaN,winter
24,10841,Georges Camille Berthet,M,20.0,NaN,NaN,France,FRA,1924,Chamonix,Military Ski Patrol,Military Ski Patrol Men's Military Ski Patrol,Bronze,winter
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
270385,129003,Marie Wattel,F,19.0,181.0,71.0,France,FRA,2016,Rio de Janeiro,Swimming,Swimming Women's 4 x 100 metres Medley Relay,NaN,summer
270695,132134,Isabelle Yacoubou-Dehoui,F,30.0,190.0,104.0,France,FRA,2016,Rio de Janeiro,Basketball,Basketball Women's Basketball,NaN,summer
270812,133031,Tony Victor James Yoka,M,24.0,200.0,105.0,France,FRA,2016,Rio de Janeiro,Boxing,Boxing Men's Super-Heavyweight,Gold,summer
270894,133577,Grce Zaadi Deuna,F,23.0,171.0,66.0,France,FRA,2016,Rio de Janeiro,Handball,Handball Women's Handball,Silver,summer


## Filtering by dates

You'll learn more about date and time in the coming week. For now, please note how the ```Year``` column is a  ```int64``` object and NOT a ```datetime``` object. We can still work with it for our needs.

In [168]:
## GET INFO

df["Year"]

0         1924
1         1924
2         1924
3         1924
4         1924
          ... 
271111    2016
271112    2016
271113    2016
271114    2016
271115    2016
Name: Year, Length: 271116, dtype: int64

In [169]:
## Find all the 1900 competitions

df[df["Year"] == 1900]

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal,Season
48944,4,Edgar Lindenau Aabye,M,34.0,NaN,NaN,Denmark/Sweden,DEN,1900,Paris,Tug-Of-War,Tug-Of-War Men's Tug-Of-War,Gold,summer
48945,133,Franz Abb,M,NaN,NaN,NaN,Germany,GER,1900,Paris,Gymnastics,Gymnastics Men's Individual All-Around,NaN,summer
48946,150,Margaret Ives Abbott (-Dunne),F,23.0,NaN,NaN,United States,USA,1900,Paris,Golf,Golf Women's Individual,Gold,summer
48947,151,Mary Perkins Ives Abbott (Perkins-),F,42.0,NaN,NaN,United States,USA,1900,Paris,Golf,Golf Women's Individual,NaN,summer
48948,404,"Louis Grenville ""Lou"" Abell",M,15.0,NaN,NaN,Vesper Boat Club,USA,1900,Paris,Rowing,Rowing Men's Coxed Eights,Gold,summer
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
50875,131733,William Hayden Wright,M,NaN,NaN,NaN,A North American Team,USA,1900,Paris,Polo,Polo Men's Polo,Bronze,summer
50876,133574,Henri Yvon,M,NaN,NaN,NaN,France,FRA,1900,Paris,Fencing,"Fencing Men's Foil, Masters, Individual",NaN,summer
50877,133574,Henri Yvon,M,NaN,NaN,NaN,France,FRA,1900,Paris,Fencing,"Fencing Men's epee, Masters, Individual",NaN,summer
50878,133807,Pyotr Antonovich Zakovorot,M,NaN,NaN,NaN,Russia,RUS,1900,Paris,Fencing,"Fencing Men's Sabre, Masters, Individual",NaN,summer


In [170]:
### return all the competitions between 1920 and 1950 (inclusive)

df.loc['1920':'1950']

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal,Season
192000,21771,douard Clarisse,M,24.0,178.0,79.0,Mauritius,MRI,1996,Atlanta,Badminton,Badminton Men's Singles,NaN,summer
192001,21771,douard Clarisse,M,24.0,178.0,79.0,Mauritius-1,MRI,1996,Atlanta,Badminton,Badminton Mixed Doubles,NaN,summer
192002,21771,douard Clarisse,M,24.0,178.0,79.0,Mauritius,MRI,1996,Atlanta,Badminton,Badminton Men's Doubles,NaN,summer
192003,21802,Gregg Clark,M,25.0,172.0,70.0,South Africa,RSA,1996,Atlanta,Hockey,Hockey Men's Hockey,NaN,summer
192004,21812,Karen Lynn Clark (-Le Poole),F,24.0,176.0,59.0,Canada,CAN,1996,Atlanta,Synchronized Swimming,Synchronized Swimming Women's Team,Silver,summer
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
194995,52762,Tetiana Ivaniushkina,F,29.0,NaN,NaN,Ukraine,UKR,1996,Atlanta,Volleyball,Volleyball Women's Volleyball,NaN,summer
194996,52777,Andrey Vladimirovich Ivanov,M,20.0,191.0,85.0,Russia,RUS,1996,Atlanta,Swimming,Swimming Men's 200 metres Breaststroke,NaN,summer
194997,52779,Anton Sashov Ivanov,M,25.0,177.0,69.0,Bulgaria,BUL,1996,Atlanta,Athletics,Athletics Men's 200 metres,NaN,summer
194998,52792,Evgeni Ivanov,M,22.0,210.0,98.0,Bulgaria,BUL,1996,Atlanta,Volleyball,Volleyball Men's Volleyball,NaN,summer


In [171]:
## Were there any games in 1950

df[df["Year"] == 1950]
## no?? 

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal,Season


In [172]:
## return only Tennis competitions between 1920 and 1950 (inclusive)

df.loc['1920':'1950'].query("Sport == 'Tennis'")

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal,Season
192054,22231,Amanda Coetzer,F,24.0,157.0,53.0,South Africa,RSA,1996,Atlanta,Tennis,Tennis Women's Singles,NaN,summer
192055,22231,Amanda Coetzer,F,24.0,157.0,53.0,South Africa,RSA,1996,Atlanta,Tennis,Tennis Women's Doubles,NaN,summer
192092,22688,Grant Connell,M,30.0,187.0,79.0,Canada,CAN,1996,Atlanta,Tennis,Tennis Men's Doubles,NaN,summer
192142,23273,"Alberto ""Albert"" Costa Casal",M,21.0,180.0,78.0,Spain,ESP,1996,Atlanta,Tennis,Tennis Men's Singles,NaN,summer
192144,23279,Carlos Costa Masferrer,M,18.0,183.0,79.0,Spain,ESP,1996,Atlanta,Tennis,Tennis Men's Singles,NaN,summer
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
194771,50696,Anke Huber,F,22.0,173.0,58.0,Germany,GER,1996,Atlanta,Tennis,Tennis Women's Singles,NaN,summer
194831,51405,Patricia Hy-Boulais,F,30.0,163.0,58.0,Canada,CAN,1996,Atlanta,Tennis,Tennis Women's Singles,NaN,summer
194832,51405,Patricia Hy-Boulais,F,30.0,163.0,58.0,Canada,CAN,1996,Atlanta,Tennis,Tennis Women's Doubles,NaN,summer
194993,52760,Goran imun Ivanievi,M,24.0,194.0,78.0,Croatia,CRO,1996,Atlanta,Tennis,Tennis Men's Singles,NaN,summer


In [173]:
## return only Women's Tennis competitions results
## between 1996 and 2018 inclusive
df.loc['1996':'2018'].query("Sport == 'Tennis'").query("Sex == 'F'")

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal,Season
199797,97255,Nicole Anne Louise Provis-Bradtke,F,26.0,176.0,61.0,Australia,AUS,1996,Atlanta,Tennis,Tennis Women's Singles,NaN,summer
199948,98846,Dally Randriantefy,F,19.0,165.0,55.0,Madagascar,MAD,1996,Atlanta,Tennis,Tennis Women's Singles,NaN,summer
199949,98846,Dally Randriantefy,F,19.0,165.0,55.0,Madagascar,MAD,1996,Atlanta,Tennis,Tennis Women's Doubles,NaN,summer
199950,98847,Natacha Randriantefy,F,18.0,173.0,52.0,Madagascar,MAD,1996,Atlanta,Tennis,Tennis Women's Doubles,NaN,summer
200326,103163,"Virginia ""Vivi"" Ruano Pascual",F,22.0,164.0,60.0,Spain,ESP,1996,Atlanta,Tennis,Tennis Women's Singles,NaN,summer
200405,104019,Gabriela Beatriz Sabatini,F,26.0,175.0,64.0,Argentina,ARG,1996,Atlanta,Tennis,Tennis Women's Singles,NaN,summer
200406,104019,Gabriela Beatriz Sabatini,F,26.0,175.0,64.0,Argentina,ARG,1996,Atlanta,Tennis,Tennis Women's Doubles,NaN,summer
200520,105256,"Arnzazu Isabel Mara ""Arantxa"" Snchez Vicario",F,24.0,169.0,56.0,Spain,ESP,1996,Atlanta,Tennis,Tennis Women's Singles,Silver,summer
200521,105256,"Arnzazu Isabel Mara ""Arantxa"" Snchez Vicario",F,24.0,169.0,56.0,Spain,ESP,1996,Atlanta,Tennis,Tennis Women's Doubles,Bronze,summer
200545,105446,Benjamas Sangaram,F,21.0,164.0,52.0,Thailand,THA,1996,Atlanta,Tennis,Tennis Women's Doubles,NaN,summer


In [174]:
## volleyball or rowing for only Male

df[(df["Sport"] == 'Volleyball') | (df["Sport"] == 'Rowing') & (df['Sex']=='M')]

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal,Season
48948,404,"Louis Grenville ""Lou"" Abell",M,15.0,NaN,NaN,Vesper Boat Club,USA,1900,Paris,Rowing,Rowing Men's Coxed Eights,Gold,summer
48963,3476,Max Heinrich Ammermann,M,21.0,NaN,NaN,Favorite Hammonia-3,GER,1900,Paris,Rowing,Rowing Men's Coxed Fours,Bronze,summer
48975,4343,Angot,M,NaN,NaN,NaN,Club Nautique de Dieppe-5,FRA,1900,Paris,Rowing,Rowing Men's Coxed Fours,NaN,summer
48980,5538,St George Ashe,M,28.0,NaN,NaN,Great Britain,GBR,1900,Paris,Rowing,Rowing Men's Single Sculls,Bronze,summer
49005,8065,Hermann Joseph Barrelet,M,20.0,NaN,NaN,France,FRA,1900,Paris,Rowing,Rowing Men's Single Sculls,Gold,summer
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
270978,134431,Zhang Changning,F,20.0,193.0,79.0,China,CHN,2016,Rio de Janeiro,Volleyball,Volleyball Women's Volleyball,Gold,summer
271023,134647,Zhao Jingbin,M,26.0,185.0,72.0,China,CHN,2016,Rio de Janeiro,Rowing,Rowing Men's Lightweight Coxless Fours,NaN,summer
271064,134857,Zhu Ting,F,21.0,198.0,78.0,China,CHN,2016,Rio de Janeiro,Volleyball,Volleyball Women's Volleyball,Gold,summer
271072,134970,Mirosaw Zitarski,M,22.0,190.0,90.0,Poland,POL,2016,Rio de Janeiro,Rowing,Rowing Men's Quadruple Sculls,NaN,summer


## What year was women's boxing introduced to the olympics?

In [175]:
## find all the names of sports first just to confirm boxing
## does not appear in different names
## how is boxing listed?

df["Sport"].unique()

<StringArray>
[               'Ice Hockey',           'Nordic Combined',
               'Ski Jumping',      'Cross Country Skiing',
                   'Curling',                 'Bobsleigh',
       'Military Ski Patrol',            'Figure Skating',
             'Speed Skating',                  'Alpinism',
                  'Skeleton',             'Alpine Skiing',
                  'Biathlon',                      'Luge',
          'Freestyle Skiing', 'Short Track Speed Skating',
              'Snowboarding',                    'Tennis',
                 'Athletics',                  'Swimming',
                'Gymnastics',                  'Shooting',
                   'Cycling',                   'Fencing',
             'Weightlifting',                 'Wrestling',
                'Tug-Of-War',                      'Golf',
                    'Rowing',                     'Rugby',
                   'Sailing',                  'Football',
                   'Cricket',             

### When did boxing first appear as a sport in which women competed at the olympics?



In [176]:
## query for boxing and women
## store in new df
boxwomen_df=df.query("Sport == 'Boxing'").query("Sex == 'F'")
boxwomen_df.sort_values (by= ["Year"], ascending = True)

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal,Season
244598,832,Nicola Virginia Adams,F,29.0,164.0,51.0,Great Britain,GBR,2012,London,Boxing,Boxing Women's Flyweight,Gold,summer
252033,79355,Karolina Michalczuk,F,32.0,164.0,51.0,Poland,POL,2012,London,Boxing,Boxing Women's Flyweight,NaN,summer
252877,87919,Sofya Albertovna Ochigava,F,25.0,170.0,60.0,Russia,RUS,2012,London,Boxing,Boxing Women's Lightweight,Silver,summer
252896,88189,Edith Ogoke,F,21.0,NaN,75.0,Nigeria,NGR,2012,London,Boxing,Boxing Women's Middleweight,NaN,summer
253072,90065,Mahjouba Oubtil,F,29.0,NaN,60.0,Morocco,MAR,2012,London,Boxing,Boxing Women's Lightweight,NaN,summer
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
263562,61647,Tetiana Volodymyrivna Kob,F,28.0,161.0,51.0,Ukraine,UKR,2016,Rio de Janeiro,Boxing,Boxing Women's Flyweight,NaN,summer
263948,65726,Hasnaa Lachgar,F,26.0,164.0,60.0,Morocco,MAR,2016,Rio de Janeiro,Boxing,Boxing Women's Lightweight,NaN,summer
264018,66595,Peamwilai Laopeam,F,32.0,163.0,51.0,Thailand,THA,2016,Rio de Janeiro,Boxing,Boxing Women's Flyweight,NaN,summer
264287,69466,Li Qian,F,26.0,180.0,75.0,China,CHN,2016,Rio de Janeiro,Boxing,Boxing Women's Middleweight,Bronze,summer


### How many values in a column?



## Counting

- The ```sum``` adds up all the values in a dataset, to give you the total amount.

- The ```count``` tells you how many items of particular values exist in a dataset. It doesn't care about the values themselves, just the number of entries.

We use the ```.value_counts()``` method:

```df_name["column_name"].value_counts()```

### Summer v. Winter

How many athletes participated in the summer games v. the winter games?

In [177]:
## number of summer athletes

summer_athletes=dfs['ID'].nunique()
(summer_athletes)

116776

In [178]:
## number of winter athletes 
winter_athletes=dfw['ID'].nunique()
(winter_athletes)

18958

In [179]:
## total athletes
115776+18958

134734

### ```value_counts()``` parameter

By default, it is ```.value_counts(normalize = False)```

This gives you want we calculated above

But ```.value_counts(normalize = True)``` returns  proportions (fractions of the total).

In [180]:
## get as proportion

total_athletes=df['ID'].value_counts(normalize=False)
(total_athletes)

ID
77710     58
106296    39
115354    38
119591    36
44875     32
          ..
135431     1
135477     1
135478     1
135525     1
135568     1
Name: count, Length: 135571, dtype: int64

In [181]:
## convert to you want it as a percentage

normalized_athletes=(total_athletes)*100

In [182]:
## subset data to hold only medals
medal_athlete=df['Medal'].value_counts(normalize=False)
(medal_athlete)

Medal
Gold      13372
Bronze    13295
Silver    13116
Name: count, dtype: int64

In [183]:
## what is the total medal count for winter v. summer?

winter_medal=dfw['Medal'].value_counts(normalize=False)
(winter_medal)

Medal
Gold      1913
Silver    1896
Bronze    1886
Name: count, dtype: int64

In [184]:
summer_medal=dfs['Medal'].value_counts(normalize=False)
(summer_medal)

Medal
Gold      11459
Bronze    11409
Silver    11220
Name: count, dtype: int64

In [185]:
total_medal=df['Medal'].value_counts(normalize=False)
(total_medal)

Medal
Gold      13372
Bronze    13295
Silver    13116
Name: count, dtype: int64

# DRY

Do not repeat yourself

In [186]:
##get list of uniques
df['Medal'].unique()

['Silver', NaN, 'Bronze', 'Gold']
Categories (3, str): ['Bronze', 'Gold', 'Silver']

In [187]:
# target_medals = ["Gold", "Silver", "Bronze"]



In [188]:
## get head


In [189]:
## call winners


In [190]:
## what is the total medal count for winter v. summer?


## How many medals were handed out at each olympics between 1896 and 2014?

Show the result from highest to lowest.

In [191]:
## how many medals were handed out at each olympics?
## show the result in descending order

In [192]:
only_medals = df[df['Medal'].notna()]
(only_medals)

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal,Season
0,391,Clarence John Abel,M,23.0,185.0,102.0,United States,USA,1924,Chamonix,Ice Hockey,Ice Hockey Men's Ice Hockey,Silver,winter
5,1341,Johan Petter hln (Andersson-),M,44.0,NaN,NaN,Sweden,SWE,1924,Chamonix,Curling,Curling Men's Curling,Silver,winter
10,3872,William Harding Anderson,M,22.0,NaN,NaN,Great Britain,GBR,1924,Chamonix,Ice Hockey,Ice Hockey Men's Ice Hockey,Bronze,winter
11,4060,Georges Jules Andr,M,47.0,NaN,NaN,France,FRA,1924,Chamonix,Curling,Curling Men's Curling,Bronze,winter
13,5232,Thomas Alfred Arnold,M,22.0,NaN,NaN,Great Britain-1,GBR,1924,Chamonix,Bobsleigh,Bobsleigh Men's Four/Five,Silver,winter
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
271064,134857,Zhu Ting,F,21.0,198.0,78.0,China,CHN,2016,Rio de Janeiro,Volleyball,Volleyball Women's Volleyball,Gold,summer
271079,135132,Bojana ivkovi,F,28.0,186.0,72.0,Serbia,SRB,2016,Rio de Janeiro,Volleyball,Volleyball Women's Volleyball,Silver,summer
271082,135205,Shakhobiddin Shokirovich Zoirov,M,23.0,169.0,52.0,Uzbekistan,UZB,2016,Rio de Janeiro,Boxing,Boxing Men's Flyweight,Gold,summer
271087,135245,Milenko Zori,M,27.0,179.0,73.0,Serbia,SRB,2016,Rio de Janeiro,Canoeing,"Canoeing Men's Kayak Doubles, 1,000 metres",Silver,summer


In [193]:
medal_counts = only_medals.groupby('Year').size().reset_index(name='Total_Medals')

In [194]:
sorted_results = medal_counts.sort_values(by='Year', ascending = False)

In [195]:
print(sorted_results)

    Year  Total_Medals
34  2016          2023
33  2014           597
32  2012          1941
31  2010           520
30  2008          2048
29  2006           526
28  2004          2001
27  2002           478
26  2000          2004
25  1998           440
24  1996          1842
23  1994           331
22  1992          2030
21  1988          1845
20  1984          1698
19  1980          1602
18  1976          1531
17  1972          1414
16  1968          1256
15  1964          1215
14  1960          1058
13  1956          1043
12  1952          1033
11  1948           987
10  1936          1025
9   1932           739
8   1928           823
7   1924           962
6   1920          1308
5   1912           941
4   1908           831
3   1906           458
2   1904           486
1   1900           604
0   1896           143


In [204]:
## top 10 count of gold medals by country

country_medal = only_medals['Team'].value_counts()
country_medal.head(10)

Team
United States    5219
Soviet Union     2451
Germany          1984
Great Britain    1673
France           1550
Italy            1527
Sweden           1434
Australia        1306
Canada           1243
Hungary          1127
Name: count, dtype: int64

In [205]:
## count of gold medals by country by percent
## only gold winner df

gold_medals = df[df['Medal'] == 'Gold']
country_gold = gold_medals['Team'].value_counts()
print(country_gold)

Team
United States    2474
Soviet Union     1058
Germany           679
Italy             535
Great Britain     519
                 ... 
Bahrain             1
Kosovo              1
Tajikistan          1
Puerto Rico         1
Singapore           1
Name: count, Length: 242, dtype: int64


In [213]:
## share of gold medals 

[(country_gold)/(country_medal)*100]

[Team
 A North American Team                  NaN
 Afghanistan                            NaN
 Algeria                          29.411765
 Ali-Baba II                            NaN
 Amateur Athletic Association    100.000000
                                    ...    
 Winnipeg Shamrocks-1            100.000000
 Yugoslavia                       33.333333
 Zambia                                 NaN
 Zimbabwe                         77.272727
 Zut                                    NaN
 Name: count, Length: 498, dtype: float64]

In [ ]:
## as percent



In [218]:
## how many times did Russia win the gold?

## sorting for just Russian athletes 
russia = df[df['NOC'] == 'RUS']
print(russia)


            ID                           Athlete Name Sex   Age  Height  \
24035      544            Sergey Mikhaylovich Abramov   M  34.0   180.0   
24100     4615                     Andrey Anufriyenko   M  23.0   178.0   
24101     4615                     Andrey Anufriyenko   M  23.0   178.0   
24102     4615                     Andrey Anufriyenko   M  23.0   178.0   
24132     6644        Igor Gayniakhmetovich Badamshin   M  27.0   175.0   
...        ...                                    ...  ..   ...     ...   
271043  134759        Viktoriya Yuryevna Zhilinskayte   F  27.0   188.0   
271044  134776  Valeriya Sergeyevna Zholobova-Koblova   F  23.0   164.0   
271107  135489  Anastasiya Valeryevna Zuyeva-Fesikova   F  26.0   182.0   
271108  135489  Anastasiya Valeryevna Zuyeva-Fesikova   F  26.0   182.0   
271109  135489  Anastasiya Valeryevna Zuyeva-Fesikova   F  26.0   182.0   

        Weight    Team  NOC  Year            City                 Sport  \
24035     94.0  Russia  

In [221]:
## sorting for Russian gold medals

russia_gold = russia[russia['Medal'] == 'Gold']
russia_gold.value_counts()

ID      Athlete Name                        Sex  Age   Height  Weight  Team      NOC  Year  City            Sport                 Event                                                Medal   Season
8967    Svetlana Valeryevna Bazhanova       F    21.0  175.0   70.0    Russia    RUS  1994  Lillehammer     Speed Skating         Speed Skating Women's 3,000 metres                   Gold    winter    1
20410   Sergey Vladimirovich Chepikov       M    27.0  182.0   75.0    Russia    RUS  1994  Lillehammer     Biathlon              Biathlon Men's 10 kilometres Sprint                  Gold    winter    1
39171   Nina Vasilyevna Gavrylyuk           F    28.0  167.0   55.0    Russia    RUS  1994  Lillehammer     Cross Country Skiing  Cross Country Skiing Women's 4 x 5 kilometres Relay  Gold    winter    1
41224   Aleksandr Vyacheslavovich Golubyov  M    21.0  174.0   74.0    Russia    RUS  1994  Lillehammer     Speed Skating         Speed Skating Men's 500 metres                       Gold    wi

In [223]:
russia_gold = df[(df['NOC'] == 'RUS') & (df['Medal'] == 'Gold')]

russia_gold


,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal,Season
24187,8967,Svetlana Valeryevna Bazhanova,F,21.0,175.0,70.0,Russia,RUS,1994,Lillehammer,Speed Skating,"Speed Skating Women's 3,000 metres",Gold,winter
24468,20410,Sergey Vladimirovich Chepikov,M,27.0,182.0,75.0,Russia,RUS,1994,Lillehammer,Biathlon,Biathlon Men's 10 kilometres Sprint,Gold,winter
24848,39171,Nina Vasilyevna Gavrylyuk,F,28.0,167.0,55.0,Russia,RUS,1994,Lillehammer,Cross Country Skiing,Cross Country Skiing Women's 4 x 5 kilometres ...,Gold,winter
24903,41224,Aleksandr Vyacheslavovich Golubyov,M,21.0,174.0,74.0,Russia,RUS,1994,Lillehammer,Speed Skating,Speed Skating Men's 500 metres,Gold,winter
24912,41785,"Yekaterina Aleksandrovna ""Katya"" Gordeyeva (-G...",F,22.0,NaN,NaN,Russia-1,RUS,1994,Lillehammer,Figure Skating,Figure Skating Mixed Pairs,Gold,winter
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
270251,127739,Anna Viktorovna Vyakhireva,F,21.0,168.0,63.0,Russia,RUS,2016,Rio de Janeiro,Handball,Handball Women's Handball,Gold,summer
270782,132795,Yana Karapetovna Yegoryan,F,22.0,175.0,64.0,Russia,RUS,2016,Rio de Janeiro,Fencing,"Fencing Women's Sabre, Individual",Gold,summer
270783,132795,Yana Karapetovna Yegoryan,F,22.0,175.0,64.0,Russia,RUS,2016,Rio de Janeiro,Fencing,"Fencing Women's Sabre, Team",Gold,summer
270792,132899,Tatyana Vladimirova Yerokhina,F,31.0,185.0,73.0,Russia,RUS,2016,Rio de Janeiro,Handball,Handball Women's Handball,Gold,summer


In [224]:
rus_codes = ['RUS', 'URS', 'EUN', 'OAR', 'ROC']
all_rus_gold = df[(df['NOC'].isin(rus_codes)) & (df['Medal'] == 'Gold')]
all_rus_gold

,ID,Athlete Name,Sex,Age,Height,Weight,Team,NOC,Year,City,Sport,Event,Medal,Season
4494,4375,Nikolay Petrovich Anikin,M,24.0,NaN,NaN,Soviet Union,URS,1956,Cortina d'Ampezzo,Cross Country Skiing,Cross Country Skiing Men's 4 x 10 kilometres R...,Gold,winter
4518,6458,Yevgeny Makarovich Babich,M,35.0,170.0,73.0,Soviet Union,URS,1956,Cortina d'Ampezzo,Ice Hockey,Ice Hockey Men's Ice Hockey,Gold,winter
4570,12493,Vsevolod Mikhaylovich Bobrov,M,33.0,180.0,79.0,Soviet Union,URS,1956,Cortina d'Ampezzo,Ice Hockey,Ice Hockey Men's Ice Hockey,Gold,winter
4893,43030,Yevgeny Romanovich Grishin,M,24.0,174.0,78.0,Soviet Union,URS,1956,Cortina d'Ampezzo,Speed Skating,Speed Skating Men's 500 metres,Gold,winter
4894,43030,Yevgeny Romanovich Grishin,M,24.0,174.0,78.0,Soviet Union,URS,1956,Cortina d'Ampezzo,Speed Skating,"Speed Skating Men's 1,500 metres",Gold,winter
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
270251,127739,Anna Viktorovna Vyakhireva,F,21.0,168.0,63.0,Russia,RUS,2016,Rio de Janeiro,Handball,Handball Women's Handball,Gold,summer
270782,132795,Yana Karapetovna Yegoryan,F,22.0,175.0,64.0,Russia,RUS,2016,Rio de Janeiro,Fencing,"Fencing Women's Sabre, Individual",Gold,summer
270783,132795,Yana Karapetovna Yegoryan,F,22.0,175.0,64.0,Russia,RUS,2016,Rio de Janeiro,Fencing,"Fencing Women's Sabre, Team",Gold,summer
270792,132899,Tatyana Vladimirova Yerokhina,F,31.0,185.0,73.0,Russia,RUS,2016,Rio de Janeiro,Handball,Handball Women's Handball,Gold,summer
